hbase-server/src/main/java/org/apache/hadoop/hbase/master/RegionsRecoveryChore.java

```diff
@@ -0,0 +1,159 @@
+   *
+   * @param stopper When {@link Stoppable#isStopped()} is true, this chore will cancel and cleanup
+   * @param configuration The configuration params to be used
+   * @param hMaster HMaster instance to initiate RegionTableRegions
+   */
+  RegionsRecoveryChore(final Stoppable stopper, final Configuration configuration,
+      final HMaster hMaster) {
+
+    super(REGIONS_RECOVERY_CHORE_NAME, stopper, configuration.getInt(REGIONS_RECOVERY_INTERVAL,
+      DEFAULT_REGIONS_RECOVERY_INTERVAL));
+    this.hMaster = hMaster;
+    this.storeRefCountThreshold = configuration.getInt(STORE_REF_COUNT_THRESHOLD,
+      DEFAULT_STORE_REF_COUNT_THRESHOLD);
+
+  }
+
+  @Override
+  protected void chore() {
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Starting up Regions Recovery by reopening regions based on storeRefCount...");
+    }
+    try {
+      final ClusterMetrics clusterMetrics = hMaster.getClusterMetrics();
+      final Map<ServerName, ServerMetrics> serverMetricsMap =
+        clusterMetrics.getLiveServerMetrics();
+      final Map<TableName, List<byte[]>> tableToReopenRegionsMap =
+        getTableToRegionsByRefCount(serverMetricsMap);
+      if (MapUtils.isNotEmpty(tableToReopenRegionsMap)) {
+        tableToReopenRegionsMap.forEach((tableName, regionNames) -> {
+          try {
+            LOG.warn("Reopening regions due to high refCount. TableName: {} , noOfRegions: {}",
+              tableName, regionNames.size());
+            hMaster.reopenRegions(tableName, regionNames, NONCE_GENERATOR.getNonceGroup(),
+              NONCE_GENERATOR.newNonce());
+          } catch (IOException e) {
+            LOG.error("{} tableName: {}, regionNames: {}", ERROR_REOPEN_REIONS_MSG,
+              tableName, regionNames, e);
+          }
+        });
+      }
+    } catch (Exception e) {
+      LOG.error("Error while reopening regions based on storeRefCount threshold", e);
+    }
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Exiting Regions Recovery by reopening regions based on storeRefCount...");
+    }
+  }
+
+  private Map<TableName, List<byte[]>> getTableToRegionsByRefCount(
+      final Map<ServerName, ServerMetrics> serverMetricsMap) {
+
+    final Map<TableName, List<byte[]>> tableToReopenRegionsMap = new HashMap<>();
+    for (ServerMetrics serverMetrics : serverMetricsMap.values()) {
+      Map<byte[], RegionMetrics> regionMetricsMap = serverMetrics.getRegionMetrics();
+      for (RegionMetrics regionMetrics : regionMetricsMap.values()) {
+        // For each region, each store file can have different ref counts
+        // We need to find maximum of all such ref counts and if that max count
+        // is beyond a threshold value, we should reopen the region.
+        // Here, we take max ref count of all stores and not the cumulative count
+        // of all stores.
+        final int maxStoreRefCount = regionMetrics.getMaxStoreRefCount();
+        if (maxStoreRefCount > storeRefCountThreshold) {
```

Comment:
We should consider the system table different?  META table might some time get too large ref count. It is possible. May be many new clients are started/restarted or many regions move/split etc.  What if at the time of reporting it had a huge ref count. Reopening of META will affect the whole system!   Anyways the def count of 256 (of now) is way less for any kind of regions IMO.

Reply:
Sure, better we exclude META? In your opinion, what default count (instead of 256) should be ideal?